### PR TITLE
Implement advanced login

### PR DESCRIPTION
**Pull request by lumosnysm**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
       log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (mdmahmudurrahman) [inline app/controllers/sessions_controller.rb]: Use Double quote

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -4,15 +4,36 @@ def log_in user
   end
 
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by id: user_id
+      if user &.authenticated? cookies[:remember_token]
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: remove redundant space

File: app/models/user.rb [hunk 3]
```diff
@@ -1,9 +1,34 @@
 class User < ApplicationRecord
+  attr_accessor :remember_token
   before_save { email.downcase! }
   validates :name,  presence: true, length: { maximum: 50 }
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   validates :email, presence: true, length: { maximum: 255 }, 
   format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, presence: true, length: { minimum: 6 }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Use config gem to place the constant in a central file
Reference: [https://github.com/railsconfig/config](https://github.com/railsconfig/config)

File: app/models/user.rb [hunk 4]
```diff
@@ -1,9 +1,34 @@
...
+
+  def User.digest string
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Use class << self bloc to declare class method

File: app/models/user.rb [hunk 5]
```diff
@@ -1,9 +1,34 @@
...
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+    BCrypt::Engine.cost
+    BCrypt::Password.create string, cost: cost
+  end
+
+  def User.new_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Use class << self bloc to declare class method

File: app/models/user.rb [hunk 6]
```diff
@@ -1,9 +1,34 @@
...
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1. Use update_attributes
2. Can you tell me the difference between update_attribute and update_attributes

Comment (lumosnysm) [reply app/models/user.rb]: I did some research.

update_attributes 
- takes an option hash of attribute names and their values.
- validate before saving.
- returns false if unable to save

update_attribute:
- takes two attributes, name and value.
- not going through the validation.
- returns false only if the query failed on the database.

File: app/models/user.rb [hunk 7]
```diff
@@ -1,9 +1,35 @@
...
+    update_attributes remember_digest: User.digest(remember_token)
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Please use update_attributes here

File: app/models/user.rb [hunk 8]
```diff
@@ -1,9 +1,35 @@
 class User < ApplicationRecord
+  attr_accessor :remember_token
```
Comment (huongnguyenmta) [inline app/models/user.rb]: **attr_accessor :remember_token** có tác dụng gì vậy em?

Comment (lumosnysm) [reply app/models/user.rb]: nó là phương thức để tự động tạo getter với setter cho :remember_token ạ.

File: app/models/user.rb [hunk 9]
```diff
@@ -1,9 +1,35 @@
...
   before_save { email.downcase! }
   validates :name,  presence: true, length: { maximum: 50 }
```
Comment (huuphuc93) [inline app/models/user.rb]: Do not add white spaces before and after [] () {}

**Pull request merged**